### DISCUSSION TITLE
X1 camera reliability

### DISCUSSION THREAD

**Discussion by sushibagel** [greghesp/ha-bambulab#453] [Q&A]
After reading up on it I know there are known issues with the way Bambu implemented the camera streaming. That being said I've yet to be able to keep the camera working for more than an hour or two at a time.

Can anyone share any tips on what they've done to keep/make the camera work?

**Comment by AdrianGarside** [comment]
The *very first* rtsp connection to the printer is far more reliable and as long as it never drops it generally keeps working.

So the best approach I've found is:
1. OPTIONAL but for best results, install go2rtc in its own docker - this keeps it running if you restart home assistant.
2. Install the webrtc integration. This includes its own go2rtc if you didn't do (1) but if you restart home assistant the rtsp feed will almost certainly break until you restart the X1
3. Add a card to view the stream configured to keep running in the background so that the rtsp feed from the X1 never stops (since it's highly likely a new connection will then not work).

My webrtc camera card config is:
```
type: custom:webrtc-camera
url: bambux1c
media: video
mode: webrtc/tcp
background: true
server: http://192.168.0.5:1984/
ui: true
style: >-
  .screenshot {display: none} .pictureinpicture {display: none} .mode {display:
  none}
```

The url can just be the full rtsp url to the X1 but I re-use this go2rtc server in my test home assistant integration so I wanted a friendly name to reference it. In my go2rtc config I place the real URL:

streams:
    bambux1c: rtspx://bblp:[ACCESS_CODE]@10.10.10.95:322/streaming/live/1

**Reply by sushibagel** [reply]
Thanks I'll give these a try

**Comment by AdrianGarside** [comment]
With this I only lose the feed about once a week or so typically.

**Reply by sushibagel** [reply]
Thanks for the response. Before I got around to doing all your suggestions I actually noticed that it seems to be the camera fails after I've viewed it in Bambu Studio or Bambu Handy. So that may be something to look into that something with activating those streams messes with the rtc stream we are using.

**Reply by AdrianGarside** [reply]
I don't have any issues viewing the video feed in studio/handy with my setup.

**Reply by sushibagel** [reply]
> I don't have any issues viewing the video feed in studio/handy with my setup.

Interesting, maybe just a coincidence. I'll report back if I continue to notice it.

**Reply by AdrianGarside** [reply]
Oh no - before I had this setup viewing the stream in the handy app pretty consistently broke the ability to load the RTSP stream. The key benefit of the above setup is you never let the first RTSP connection drop so that you never have to connect fresh since that breaks pretty quickly.

**Comment by capsel22** [comment]
> The _very first_ rtsp connection to the printer is far more reliable and as long as it never drops it generally keeps working.
> 
> So the best approach I've found is:
> 
>     1. OPTIONAL but for best results, install go2rtc in its own docker - this keeps it running if you restart home assistant.
> 
>     2. Install the webrtc integration. This includes its own go2rtc if you didn't do (1) but if you restart home assistant the rtsp feed will almost certainly break until you restart the X1
> 
>     3. Add a card to view the stream configured to keep running in the background so that the rtsp feed from the X1 never stops (since it's highly likely a new connection will then not work).
> 
> 
> My webrtc camera card config is:
> 
> ```
> type: custom:webrtc-camera
> url: bambux1c
> media: video
> mode: webrtc/tcp
> background: true
> server: http://192.168.0.5:1984/
> ui: true
> style: >-
>   .screenshot {display: none} .pictureinpicture {display: none} .mode {display:
>   none}
> ```
> 
> The url can just be the full rtsp url to the X1 but I re-use this go2rtc server in my test home assistant integration so I wanted a friendly name to reference it. In my go2rtc config I place the real URL:
> 
> streams: bambux1c: rtspx://bblp:[ACCESS_CODE]@10.10.10.95:322/streaming/live/1

Stupid question but, if I use custom-webrtc card, I don't have preload option, only if I use the HA's default Picture Glance card. is this correct? Or do you mean the go2rtc server will keep the feed alive instead

**Comment by AdrianGarside** [comment]
background: true
Keeps it alive once viewed and that’s critical for keeping the feed working. My experience is that if that connection ever drops (e.g. the HA server restarts) then I’ll have to reboot the X1C to get a working connection again.

**Reply by capsel22** [reply]
Yes, make sense. I actually added it to go2rtc as a custom feed, piped that link into generic camera entity in HA and used default picture glance with "preload stream" pointed at the generic camera. Seems to be working quite well. Time will tell

**Comment by metawops** [comment]
I, too, am pretty annoyed by this unreliability of the X1C's video feed.
I will try to understand the suggestions from this thread.
But I'm wondering: Are the problems solely on Bambu Lab's side or ist there a way for the developer of this integration to permanently and successfully work around the shortcomings in his code (and he "just" needs to do it)? 😳

**Reply by AdrianGarside** [reply]
Totally on the Bambu side. There is nothing we can do to make it better in the integration. There is a way you can configure a different video re-feed integration to make it more reliable as the very first connection after the printer is turned on is much more reliable - you just need to not ever drop that connection. Even then it is still buggy.

**Reply by sophof** [reply]
I checked bambu's forum, but couldn't find anything there connected to this issue. So I don't think Bambu is even aware of this bug? I'm not knowledgeable enough about it to make my own report, but if someone here is willing that might at least fix the issue there?
